### PR TITLE
Drivetrain

### PR DESCRIPTION
**Pull request by CoolSpy3**

### CODE REVIEW

File: src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Drivetrain.java
Comment (brettle) [inline src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Drivetrain.java]: The last parameter passed to `addVisionMeasurement()` here is incorrect. See here: https://docs.limelightvision.io/en/latest/apriltags_in_3d.html#using-wpilib-s-pose-estimator

**Review by brettle — CHANGES_REQUESTED** [review]

File: src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Drivetrain.java
Comment (CoolSpy3) [reply src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Drivetrain.java]: My bad; I missed that! Should be fixed :D

File: src/main/deploy/pathplanner/135.path
Comment (ProfessorAtomicManiac) [inline src/main/deploy/pathplanner/135.path:1]: I heard that Uday worked on paths. Are these the exact files that he worked on?

File: src/main/java/org/carlmontrobotics/robotcode2023/Constants.java
Comment (ProfessorAtomicManiac) [inline src/main/java/org/carlmontrobotics/robotcode2023/Constants.java]: add units for future reference?

File: src/main/deploy/pathplanner/135.path
Comment (CoolSpy3) [reply src/main/deploy/pathplanner/135.path:1]: I don't know, @Knight4 ?

Comment (CoolSpy3) [reply src/main/deploy/pathplanner/135.path:1]: You can also try pinging him on Mattermost if he doesn't respond here

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/AlignChargingStation.java
Comment (ProfessorAtomicManiac) [inline src/main/java/org/carlmontrobotics/robotcode2023/commands/AlignChargingStation.java:53]: When we were testing autopaths, if the wheels were not turned, the robot will slide when the charging station is unbalanced even in brake mode. We could potentially add something here to turn the wheels so the robot wouldn't slide.

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/CorrectToPoint.java
Comment (ProfessorAtomicManiac) [inline src/main/java/org/carlmontrobotics/robotcode2023/commands/CorrectToPoint.java]: this. although easy fix would just to use "useContinuousInput" on pid controller

File: src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Drivetrain.java
Comment (ProfessorAtomicManiac) [inline src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Drivetrain.java]: you can remove this. This was just me testing whether the drivetrain can drive at a constant angular velocity (which it didn't). There probably is a method in the kinematics class that does this.

**Review by ProfessorAtomicManiac — COMMENTED** [review]
I haven't really looked at ports since Uday has just posted them and he's missing some buttons, but overall I only commented on minor issues and its ready to be merged.

File: Webots/controllers/DeepBlueSim/DeepBlueSim.jar
Comment (brettle) [inline Webots/controllers/DeepBlueSim/DeepBlueSim.jar:1]: Is this supposed to be committed or ignored?

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/AlignChargingStation.java
Comment (brettle) [inline src/main/java/org/carlmontrobotics/robotcode2023/commands/AlignChargingStation.java]: There is a bug on both of these lines.

File: src/main/java/org/carlmontrobotics/robotcode2023/Robot.java
Comment (brettle) [inline src/main/java/org/carlmontrobotics/robotcode2023/Robot.java]: I'm not sure that is the best value this year given how scoring works.

File: src/main/java/org/carlmontrobotics/robotcode2023/Constants.java
Comment (brettle) [inline src/main/java/org/carlmontrobotics/robotcode2023/Constants.java]: What are the units for tolerance? If it's meters, that seems rather large.

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/RotateToFieldRelativeAngle.java
Comment (brettle) [inline src/main/java/org/carlmontrobotics/robotcode2023/commands/RotateToFieldRelativeAngle.java]: Doesn't this mean that the driver can't control the forward/strafe of the robot while this command is executing?

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/RotateToFieldRelativeAngle.java
Comment (brettle) [inline src/main/java/org/carlmontrobotics/robotcode2023/commands/RotateToFieldRelativeAngle.java:38]: What will happen if the robot still has angular velocity when it reaches the target angle?

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/TeleopDrive.java
Comment (brettle) [inline src/main/java/org/carlmontrobotics/robotcode2023/commands/TeleopDrive.java]: Fix formatting.

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/TeleopDrive.java
Comment (brettle) [inline src/main/java/org/carlmontrobotics/robotcode2023/commands/TeleopDrive.java]: Where did this value come from?

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/TeleopDrive.java
Comment (brettle) [inline src/main/java/org/carlmontrobotics/robotcode2023/commands/TeleopDrive.java]: Units don't match.

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/TeleopDrive.java
Comment (brettle) [inline src/main/java/org/carlmontrobotics/robotcode2023/commands/TeleopDrive.java]: Units don't match.

**Review by brettle — CHANGES_REQUESTED** [review]

File: Webots/controllers/DeepBlueSim/DeepBlueSim.jar
Comment (CoolSpy3) [reply Webots/controllers/DeepBlueSim/DeepBlueSim.jar:1]: It should be committed, otherwise the packaged `wbt` file points to a non-existent controller. In a separate note, I'm pretty sure that file is not up-to-date with the current robot, given that we haven't been using simulation, can we continue with this PR as-is, or should it be held until we can get around to updating the model?

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/AlignChargingStation.java
Comment (CoolSpy3) [reply src/main/java/org/carlmontrobotics/robotcode2023/commands/AlignChargingStation.java]: Only bug I can see is that I'm pretty sure the "front" of the robot changed between when I wrote this code and now. I can't update that until I get a chance to see the new orientation of the gyro, but is there something else I missed?

File: src/main/java/org/carlmontrobotics/robotcode2023/Robot.java
Comment (CoolSpy3) [reply src/main/java/org/carlmontrobotics/robotcode2023/Robot.java]: Updated to 5000ms which should be above the 3s used for charge-station scoring

File: src/main/java/org/carlmontrobotics/robotcode2023/Constants.java
Comment (CoolSpy3) [reply src/main/java/org/carlmontrobotics/robotcode2023/Constants.java]: Units should be m/s, the values have been updated.

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/RotateToFieldRelativeAngle.java
Comment (CoolSpy3) [reply src/main/java/org/carlmontrobotics/robotcode2023/commands/RotateToFieldRelativeAngle.java]: The idea is correct, the calls to teleopDrive should allow the driver to control forward and strafe. I'll agree that this usage, however, is incorrect, currentForward and currentStrafe are for teleopDrive internal use only. I've updated the code to use the correct method and privated those variables.

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/RotateToFieldRelativeAngle.java
Comment (CoolSpy3) [reply src/main/java/org/carlmontrobotics/robotcode2023/commands/RotateToFieldRelativeAngle.java:38]: AFAIK, `atSetpoint` also checks the robot velocity, I've added code to set a velocity setpoint for this purpose using the same values as for `CorrectToPoint`.

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/TeleopDrive.java
Comment (CoolSpy3) [reply src/main/java/org/carlmontrobotics/robotcode2023/commands/TeleopDrive.java]: Idk, I've updated it to the robot period, we should retest to make sure the controls still have good gains.

**Comment by CoolSpy3** [comment]
Comments should be addressed, I'll re-request reviews. If possible, I'd like to also retest teleop drive and charge station align to make sure nothing got broken.

File: Webots/controllers/DeepBlueSim/DeepBlueSim.jar
Comment (brettle) [reply Webots/controllers/DeepBlueSim/DeepBlueSim.jar:1]: Ok. Don't let an old model hold up this PR.

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/AlignChargingStation.java
Comment (brettle) [reply src/main/java/org/carlmontrobotics/robotcode2023/commands/AlignChargingStation.java]: Yes. Think about the range of values for `chargeStationAlignFF`.

File: src/main/java/org/carlmontrobotics/robotcode2023/Constants.java
Comment (brettle) [inline src/main/java/org/carlmontrobotics/robotcode2023/Constants.java:142]: The wheels can just spin in place. I don't think this should be needed, but I don't object.

**Review by brettle — DISMISSED** [review]

**Review by brettle — CHANGES_REQUESTED** [review]
chargeStationAlignFF issue needs to be addressed.

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/AlignChargingStation.java
Comment (CoolSpy3) [reply src/main/java/org/carlmontrobotics/robotcode2023/commands/AlignChargingStation.java]: Ah....

File: src/main/java/org/carlmontrobotics/robotcode2023/Constants.java
Comment (CoolSpy3) [reply src/main/java/org/carlmontrobotics/robotcode2023/Constants.java:142]: Agree that it's not needed, but `drive` already handles all the necessary kinematics/wheel optimization/etc. Better is the enemy of good enough (especially when we don't have to maintain 2 drive pipelines). In this case I think it should be fine. The code calls `dt.stop()` immediately after, so drive speed should (ultimately) get set to zero. This is just to get around the "don't move if speed == 0" line.

**Comment by CoolSpy3** [comment]
Alright, at this point, I think the code should be good for review. Even if it passes, I still want to conduct the aforementioned tests before this PR is merged.

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/AlignChargingStation.java
Comment (brettle) [inline src/main/java/org/carlmontrobotics/robotcode2023/commands/AlignChargingStation.java]: This might "work" but at a minimum `chargeStationAlignSpeed` and `chargeStationAlignFF` should be renamed because they don't have their normal units of m/s and volts respectively.

Even better, use the drivetrain's `SimpleFeedForward` model to compute the voltage required to overcome the tilt and move at the desired velocity.

**Review by brettle — CHANGES_REQUESTED** [review]

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/AlignChargingStation.java
Comment (CoolSpy3) [reply src/main/java/org/carlmontrobotics/robotcode2023/commands/AlignChargingStation.java]: I've updated the variable names to include units. As for using a Feedforward model, atm I don't think we have FF constants for the drivetrain as a whole, just the individual wheels. Also, those don't account for gravity, instead assuming that it will always be canceled out by the normal force.

**Comment by CoolSpy3** [comment]
My bad, review requested on wrong PR.

**Review by brettle — APPROVED** [review]

File: src/main/java/org/carlmontrobotics/robotcode2023/commands/AlignChargingStation.java
Comment (brettle) [reply src/main/java/org/carlmontrobotics/robotcode2023/commands/AlignChargingStation.java]: To keep things simple I'm fine with the current approach for now.

Fwiw, if you wanted to test what I was thinking of, you'd set the `pitchSupplier` and `rollSupplier` passed to the `SwerveModule` constructor to return the gyro's current pitch and roll (instead of just the initial pitch and roll they currently return) and set `chargeStationAlignSpeedMpSPerDeg` to 0. Basically `SwerveModule` already has the anti-gravity code. We're just not using it at the moment.

Comment (CoolSpy3) [reply src/main/java/org/carlmontrobotics/robotcode2023/commands/AlignChargingStation.java]: Ah, yes I remember proposing that in the initial versions of the command in addition to what we have now. Can't recall why we ended up with this solution. We should maybe consider it for future revisions, but atm I'm inclined to keep it as-is.

**Comment by CoolSpy3** [comment]
Marking this PR as a draft until aforementioned testing is complete...

**Review by ProfessorAtomicManiac — APPROVED** [review]

**Pull request merged**
